---
title: "Summary of CNVs called with PennCNV"
author: "Eilis Hannon"
date: "31st Jan 2020"
output: html_document
---

```{r, echo = FALSE}
#args<-commandArgs(trailingOnly = TRUE)
#fileName<-args[1]
#superPop<-args[2]
#folder<-dirname(fileName)

findOverlapsMinProp<-function(query, subject, pThres){
	# function to find overlap between two Granges based on minimum proportion
	hits <- findOverlaps(query, subject)
	overlaps <- pintersect(query[queryHits(hits)], subject[subjectHits(hits)])
	percentOverlap <- width(overlaps) / width(subject[subjectHits(hits)])
	hits <- hits[percentOverlap > pThres]
	hitsOut <- query[queryHits(hits)]
	mcols(hitsOut)$Locus<-subject$Locus[subjectHits(hits)]
	mcols(hitsOut)$hg38<-subject$hg38[subjectHits(hits)]
	return(hitsOut)
}


library(GenomicRanges)
library("chromoMap")

setwd("/gpfs/mrc0/projects/Research_Project-MRC190311/SNPdata/CNV/")
dat<-read.table("PennCNVOutput/SCZ_GCModel_MergedFiltered_AnnoGencodev29.rawcnv", stringsAsFactors=FALSE)
colnames(dat)<-c("Coordinates", "numsnp", "length", "type", "IndividualID", "startsnp", "endsnp", "genes", "V9")
dat$numsnp<-as.numeric(gsub("numsnp=", "",dat$numsnp))
dat$length<-as.numeric(gsub(",", "", gsub("length=", "",dat$length)))


## separate coordnates into chr,bp
bprange<-unlist(lapply(strsplit(dat$Coordinates, ":"), tail, n = 1))

dat<-cbind(unlist(lapply(strsplit(dat$Coordinates, ":"), head, n = 1)),
as.numeric(unlist(lapply(strsplit(bprange, "-"), head, n = 1))),
as.numeric(unlist(lapply(strsplit(bprange, "-"), tail, n = 1))),
dat)
colnames(dat)[c(1:3)]<-c("chr", "start", "end")
```

## Summary of CNVs Called

In total we identified `{r nrow(dat)}` CNVs across `r length(unique(dat$IndividualID))` individuals. The mean number of CNVs per person was `r signif(mean(as.numeric(table(dat$IndividualID))),3)` (SD = `r signif(sd(as.numeric(table(dat$IndividualID))),3)`).

```{r}

hist(as.numeric(table(dat$IndividualID)), xlab = "number of CNVs", ylab = "number of individuals", main = "")

```

CNVs were a mean of `r signif(dat$length,3)` bp (SD = `r signif(sd(dat$length),3)` bp). 

```{r}
par(mfrow = c(1,2))
hist(dat$numsnp, xlab = "number of markers", ylab = "number of CNVs", breaks = 35, main = "All")
hist(dat$length/1000, xlab = "length (kb)", ylab = "number of CNVs", breaks = 35, main = "All")
index.del<-which(dat$type == "state2,cn=1")
index.dup<-which(dat$type == "state5,cn=3")
```

`r length(index.del)` CNVs were deletions and r length(index.dup)` CNVs were duplications.

```{r}
hist(dat$numsnp[index.del], xlab = "number of markers", breaks = 35, main = "Deletions")
hist(dat$length[index.del]/1000, xlab = "length (kb)", breaks = 35, main = "Deletions")

hist(dat$numsnp[index.dup], xlab = "number of markers", breaks = 35, main = "Duplications")
hist(dat$length[index.dup]/1000, xlab = "length (kb)", breaks = 35, main = "Duplications")

```

## Summarize CNV by person

Below are histograms for the aggregation of CNV burden per person

```{r}
## summarise CNVs by person
par(mfrow = c(1,2))
totByPerson<-aggregate(dat$length/1000, by = list(dat$IndividualID), sum)
muByPerson<-aggregate(dat$length/1000, by = list(dat$IndividualID), mean)
hist(totByPerson[,2], xlab = "Combined length of CNVs (kb)", ylab = "Number of individuals", breaks = 35, main = "All")
hist(muByPerson[,2], xlab = "Mean length of CNVs (kb)", ylab = "Number of individuals", breaks = 35, main = "All")
totByPerson<-aggregate(dat$length[index.del]/1000, by = list(dat$IndividualID[index.del]), sum)
muByPerson<-aggregate(dat$length[index.del]/1000, by = list(dat$IndividualID[index.del]), mean)
hist(totByPerson[,2], xlab = "Combined length of CNVs (kb)", ylab = "Number of individuals", breaks = 35, main = "Deletions")
hist(muByPerson[,2], xlab = "Mean length of CNVs (kb)", ylab = "Number of individuals", breaks = 35, main = "Deletions")
totByPerson<-aggregate(dat$length[index.dup]/1000, by = list(dat$IndividualID[index.dup]), sum)
muByPerson<-aggregate(dat$length[index.dup]/1000, by = list(dat$IndividualID[index.dup]), mean)
hist(totByPerson[,2], xlab = "Combined length of CNVs (kb)", ylab = "Number of individuals", breaks = 35, main = "Duplications")
hist(muByPerson[,2], xlab = "Mean length of CNVs (kb)", ylab = "Number of individuals", breaks = 35, main = "Duplications")



```

## Genomic Location

Here we will plot the chromosomal distribution of CNVs.

```{r}
barplot(table(as.numeric(gsub("chr", "", dat$chr))))
```

```{r}
library("karyoploteR")

#cbind(paste("CNV",rownames(dat), sep = ""), dat$chr, dat$start, dat$end)

plot.params <- getDefaultPlotParams(plot.type=2)
plot.params$ideogramheight <- 50
plot.params$data1height <- 400
plot.params$data1inmargin <- 1
plot.params$data2height <- 0
plot.params$data1max<-33

plotDefaultPlotParams(plot.type = 2, plot.params=plot.params)

kp <- plotKaryotype(genome="hg38", plot.type=2,chromosomes=c("autosomal"), plot.params=plot.params)
kpDataBackground(kp, data.panel = 1)
kpAxis(kp, ymin=0, ymax=max(mcols(cov.del)[,1]), col="gray50", cex=0.5)
kpBars(kp, cov.del, col = "red", border = "red", ymax = max(mcols(cov.del)[,1])

kp <- plotKaryotype(genome="hg38", plot.type=2,chromosomes=c("autosomal"), plot.params=plot.params)
kpDataBackground(kp, data.panel = 1)
#kpAxis(kp, ymin=0, ymax=max(mcols(cov.del)[,1]), col="gray50", cex=0.5)
kpHeatmap(kp, cov.del, col = c("white","lightblue", "navy"), ymax = 10)


```

## Comparision with Lists of Pathogenic CNV


```{r}
index<-which(dat$type == "state2,cn=1")
allCNVs.del<-GRanges(dat$Coordinates[index])
mcols(allCNVs.del)$SampleID <- dat$IndividualID[index]
mcols(allCNVs.del)$Type <- dat$type[index]
index<-which(dat$type == "state5,cn=3")
allCNVs.dup<-GRanges(dat$Coordinates[index])
mcols(allCNVs.dup)$SampleID <- dat$IndividualID[index]
mcols(allCNVs.dup)$Type <- dat$type[index]
pThres<-0.9 ## set as minimum overlap required

## count number of CNVs at each position
cov.del<-as(coverage(allCNVs.del), "GRanges")
names(mcols(cov.del))<-"y"
cov.dup<-as(coverage(allCNVs.dup), "GRanges")
names(mcols(cov.dup))<-"y"

## do any overlap known scz cnv loci ## list taken from Rees et al. Br J Psychiatry (merge tables 1 & 2)
sczLoci<-read.csv("../../References/CNV/SCZ_CNVloci.csv", skip = 1, stringsAsFactors = FALSE)
## filter to those significant in MetaAnalysis
sczLoci<-sczLoci[which(sczLoci$significantMeta == "*"),]
sczLoci.hg38.del<-GRanges(sczLoci$hg38[grep("del", sczLoci$Locus)])
mcols(sczLoci.hg38.del)$Locus<-sczLoci$Locus[grep("del", sczLoci$Locus)]
mcols(sczLoci.hg38.del)$hg38<-sczLoci$hg38[grep("del", sczLoci$Locus)]
sczLoci.hg38.dup<-GRanges(sczLoci$hg38[grep("dup", sczLoci$Locus)])
mcols(sczLoci.hg38.dup)$Locus<-sczLoci$Locus[grep("dup", sczLoci$Locus)]
mcols(sczLoci.hg38.dup)$hg38<-sczLoci$hg38[grep("dup", sczLoci$Locus)]


overlapDel<-findOverlapsMinProp(allCNVs.del, sczLoci.hg38.del, pThres)
overlapDup<-findOverlapsMinProp(allCNVs.dup, sczLoci.hg38.dup, pThres)
output<-rbind(data.frame(overlapDel), data.frame(overlapDup))
write.csv(output, "CNVsoverlappingKnownSCZRiskLoci.csv")
```

First, we will consider identified as associated with schizophrenia. In this sample `r nrow(output)` individuals had a CNV overlapping the list of `r nrow(sczLoci)` CNVs. To be classied as overlapping, `r pThres*100`% bp of the CNV had to overlap the known loci.

```{r}
## do any overlap known ID cnv loci ## list taken from Rees et al. JAMA Psychiatry 2016 (eTable 2)
idLoci<-read.csv("../../References/CNV/IDCNVLoci.csv", stringsAsFactors = FALSE)
idLoci<-idLoci[which(idLoci$hg38 != ""),] ## 1 region I couldn't lift over
idLoci.hg38.del<-GRanges(idLoci$hg38[grep("del", idLoci$Syndrome)])
mcols(idLoci.hg38.del)$Locus<-idLoci$Syndrome[grep("del", idLoci$Syndrome)]
mcols(idLoci.hg38.del)$hg38<-idLoci$hg38[grep("del", idLoci$Syndrome)]
idLoci.hg38.dup<-GRanges(idLoci$hg38[grep("dup", idLoci$Syndrome)])
mcols(idLoci.hg38.dup)$Locus<-idLoci$Syndrome[grep("dup", idLoci$Syndrome)]
mcols(idLoci.hg38.dup)$hg38<-idLoci$hg38[grep("dup", idLoci$Syndrome)]

overlapDel<-findOverlapsMinProp(allCNVs.del, idLoci.hg38.del, pThres)
overlapDup<-findOverlapsMinProp(allCNVs.dup, idLoci.hg38.dup, pThres)
output<-rbind(data.frame(overlapDel), data.frame(overlapDup))
write.csv(output, "CNVsoverlappingIDRiskLoci.csv")
```

Next, we will consider CNVs identified as associated with intellectual disability. In this sample `r nrow(output)` individuals had a CNV overlapping the list of `r nrow(idLoci)` CNVs. 

```{r}
## do any overlap known pathogenic cnv loci ## list taken from Kendall et al 2017 Biol Psychiatry
pathLoci<-read.table("../../References/CNV/PathogenicCNVLoci.txt", stringsAsFactors = FALSE, sep = "\t", header = TRUE)
pathLoci<-pathLoci[which(pathLoci$hg38 != ""),] ## 1 region I couldn't lift over
pathLoci.hg38.del<-GRanges(pathLoci$hg38[grep("del", pathLoci$CNV.locus)])
mcols(pathLoci.hg38.del)$Locus<-pathLoci$CNV.locus[grep("del", pathLoci$CNV.locus)]
mcols(pathLoci.hg38.del)$hg38<-pathLoci$hg38[grep("del", pathLoci$CNV.locus)]
pathLoci.hg38.dup<-GRanges(pathLoci$hg38[grep("dup", pathLoci$CNV.locus)])
mcols(pathLoci.hg38.dup)$Locus<-pathLoci$CNV.locus[grep("dup", pathLoci$CNV.locus)]
mcols(pathLoci.hg38.dup)$hg38<-pathLoci$hg38[grep("dup", pathLoci$CNV.locus)]

pThres<-0.9 ## set as minimum overlap required
overlapDel<-findOverlapsMinProp(allCNVs.del, pathLoci.hg38.del, pThres)
overlapDup<-findOverlapsMinProp(allCNVs.dup, pathLoci.hg38.dup, pThres)
output<-rbind(data.frame(overlapDel), data.frame(overlapDup))
write.csv(output, "CNVsoverlappingPathogenicCNV.csv")

## do any overlap pathogenic cnv loci

```

Finally, we will consider pathogenic CNV loci. In this sample `r nrow(output)` individuals had a CNV overlapping the list of `r nrow(pathLoci)` CNVs. 

