---
title: 'ATACSeq Data QC Report: Pilot'
author: "Eilis Hannon"
date: "`r Sys.Date()`"
output: html_document
params:
  configFile: config.r
---

```{r setup, echo = FALSE, message = FALSE, warning = FALSE,results='hide'}

source(configFile) ## contains paths to files

library(ChIPQC)
library(knitr)
library(dplyr)
library(RColorBrewer)
library(ATACseqQC)
#library(BiocParallel)
#library(org.Hs.eg.db)
#library(GenomeInfoDb)
library(ChIPseeker)
library(vioplot)
library(kableExtra)
library(ChIPpeakAnno)

```

```{r loadDat, echo = FALSE, message = FALSE,results='hide'}
setwd(dataDir)
## as sample sheet is date stamped this might not work:
sampleSheet<-read.csv(paste0("SampleSheetForATACSeqQC", format(Sys.time(), "%d%b%Y"), ".csv"), stringsAsFactors = FALSE, row.names = 1)

## check if data already loaded and saved
if(file.exists(paste0(setwd(dataDir),"/QCOutput/ATACSeqQCObject_1.rdata"))){
	load("ATACSeqQCObject_1.rdata")
} else{
	print("QC data  1 not found")
}
setwd(dataDir)
if(file.exists(paste0(setwd(dataDir),"/QCOutput/ATACSeqQCObject_2.rdata"))){
	load("ATACSeqQCObject_2.rdata")
} else{
	print("QC data  2 not found")
}

fileNames<-colnames(hist.data)[-1]

lineCol<-brewer.pal(length(fileNames), "Set1")
setwd(alignedDir)

folders<-unique(sampleSheet$DataFolder)
mqcFileNames<-NULL
for(entry in folders){
  mqcFileNames<-c(mqcFileNames ,list.files(paste0(dataDir, "/",entry), pattern = "multiqc_fastqc.txt", recursive = TRUE, full.names = TRUE))
}
fastqcStats<-NULL
for(entry in mqcFileNames){
  tmp<-read.table(entry, sep = "\t", header = TRUE)
  if(!is.null(fastqcStats) && ncol(tmp) != ncol(fastqcStats)){
    sharedCols<-intersect(colnames(tmp), colnames(fastqcStats))
    tmp<-tmp[,sharedCols]
    fastqcStats<-fastqcStats[,sharedCols]
  }
  fastqcStats<-rbind(fastqcStats, tmp)
}
fastqcStats<-unique(fastqcStats)


bowtieFileNames<-NULL
for(entry in folders){
  bowtieFileNames<-c(bowtieFileNames ,list.files(paste0(alignedDir, entry), pattern = "multiqc_bowtie2.txt", recursive = TRUE, full.names = TRUE))
}
bowtieStats<-NULL
for(entry in bowtieFileNames){
  tmp<-read.table(entry, sep = "\t", header = TRUE)
  if(!is.null(bowtieStats) && ncol(tmp) != ncol(bowtieStats)){
    sharedCols<-intersect(colnames(tmp), colnames(bowtieStats))
    tmp<-tmp[,sharedCols]
    bowtieStats<-bowtieStats[,sharedCols]
  }
  bowtieStats<-rbind(bowtieStats, tmp)
}


##load summary of reads per chr in bam file
bamStats<-lapply(sub("_depDup_q30.bam", "_statsperchr.txt", sampleSheet$bamReads), read.table)

## load ENCODE QC metrics
eqcFileNames<-list.files(pattern="pbc.qc", recursive = TRUE)
eqcFileNames<-eqcFileNames[unlist(sapply(sampleSheet$SampleID, grep, eqcFileNames))]
endcodeQC<-matrix(data = unlist(lapply(eqcFileNames, read.table)), ncol = 7, byrow = TRUE)

```

This report provides a summary of an ATAC-seq experiement from the resulting sequencing data. It looks at metrics from the raw sequencing data, alignment, filtering and peak calling steps. The bioinformatic pipeline parallels that of the [ENCODE ATAC-seq guidelines] (https://www.encodeproject.org/atac-seq/). Raw fastq files have been have been aligned with Bowtie2 to the human genome (hg38) and peaks called with MACS2 for each individual sample. As this is pilot data the pipeline is still under development.

This report makes use of the quality control metrics generated as part of that pipeline, 2 R packages [ATACseqQC R package](https://bioconductor.org/packages/release/bioc/html/ATACseqQC.html) and [ChIPQC R package](http://bioconductor.org/packages/release/bioc/html/ChIPQC.html) as well as other tools such as [BBMap](https://sourceforge.net/projects/bbmap/) bbcountunique.sh tool. 

## Overview

To provide an overall summary here is a table of the number of raw reads, aligned reads (post filtering) and peaks called for each sample.

```{r}
r1_index<-match(paste0(sampleSheet$SampleID, "_r1"), fastqcStats$Sample)
r2_index<-match(paste0(sampleSheet$SampleID, "_r2"), fastqcStats$Sample)
stats.tmp<-cbind(fastqcStats[r1_index, "Total.Sequences"], fastqcStats[r2_index, "Total.Sequences"], reads(dat),unlist(lapply(peaks(dat), length)))

colnames(stats.tmp)<-c("Reads (R1)", "Reads (R2)", "Filtered Aligned Reads", "Peaks")

kable(stats.tmp)


```

## 1. Sequencing

### 1.1 Data summary


```{r, results = "asis", echo = FALSE, message = FALSE}


## get stats for r1

r1_dat<-fastqcStats[r1_index, c("Total.Sequences", "avg_sequence_length", "Sequence.length", "X.GC", "total_deduplicated_percentage")]

colnames(r1_dat)<-c("nReads", "Mean(length)", "Range(length)", "%GC", "%Distinct")

## if paired end get stats for r2
if(sum(sampleSheet$ReadType == "Paired") > 0){
	r2_dat<-fastqcStats[r2_index,c("Total.Sequences", "avg_sequence_length", "Sequence.length", "X.GC", "total_deduplicated_percentage")]
}
colnames(r2_dat)<-c("nReads", "Mean(length)", "Range(length)", "%GC", "%Distinct")



kable(cbind(r1_dat, r2_dat), digits = 2, row.names = FALSE) %>%
  kable_styling(full_width = F, position = "center") %>%
  	add_header_above(header = setNames(c(ncol(r1_dat),ncol(r2_dat)),c("Read 1","Read 2")))

```

In total there were `r sum(c(r1_dat[,1], r2_dat[,1]))/1000000` million reads generated across these samples, with a mean of 
`r signif(mean(c(r1_dat[,1], r2_dat[,1]))/1000000,3)` million per sample (SD = `r signif(sd(c(r1_dat[,1], r2_dat[,1]))/1000000,3)` million).


## 1.2 Estimate Library Complexity

A good ATAC-seq experiment will capture enrichment at a broad range of genomic locations. The number of distinct or unique reads in an indicator of the diversity or complexity of the data. If the sample does not contain sufficient DNA and/or you over-sequence, you will simply be repeatedly sequencing PCR duplicates of a restricted pool of distinct DNA fragments. This is known a low-complexity library and is not desirable.  

Complexity can be assesed at two points from both the raw sequencing reads and the aligned reads. In this section we will assess the number of unique reads as the total number of reads increases. This is equivalent to subsampling the data. The dynamics of the resulting curve indicate the complexity of the library prep and are best interpreted across a number of samples. 

```{r, plotUniquenessCurve, echo = FALSE, fig.height = 6, fig.width = 8, fig.caption = "Figure 1: Line graph of number of unique reads (y-axis) as number of reads (x-axis) increases.", message = FALSE,results='hide', fig.align = "center"}
## plot uniqueness curves
## calc percentage duplicates as sequencing increases
perc.dup<-(hist.data[,1] - hist.data[,-1])/hist.data[,1]*100

par(mar = c(5,5,1,1))
x_lim<-c(0,max(hist.data$count))
y_lim<-c(0, max(hist.data[,-1], na.rm = TRUE))
plot(hist.data$count, hist.data$count, type = "n", xlab = "nMillionReads", ylab = "nMillionUniqueReads", xlim = x_lim, ylim = y_lim, axes = FALSE)
for(i in 2:ncol(hist.data)){
	lines(hist.data$count, hist.data[,i], col = lineCol[i-1],  lwd = 1.5)
}
box()
axis(1, at = seq(0,1.5e8, 2e7), seq(0,1.5e8, 2e7)/1000000)
axis(2, at = seq(0,1.5e8, 2e6), seq(0,1.5e8, 2e6)/1000000)
legend("bottomright", colnames(hist.data)[-1], lty = 1, col = lineCol, lwd = 2)

```

To characterise the curve for each sample, we fitted a [monod function](https://en.wikipedia.org/wiki/Monod_equation);  and compared the estimated parameters. For reference, the monod function was initially developed to model the growth of microorganisms but is applicable in any situtation where a system is moving towards saturation and the rate of expansion is increasingly restricted by available resources. In this application, there is a finite number of unique reads (indicative of library complexity), this is indicated by the assymptote of the Monod function.  While, ideally, we would like to capture the complete complexity of a sample, there will come a point where additional sequencing will become inefficient (as the chance of obtaining  the reads we have yet to measure is increasingly small). Calculating the assymptote (i.e. the maximum number of unique reads) enables comparision of complexity. Lower assymptote values indicate a less complex library. Progress along the fitted function enables assessment of sequencing sufficiency. The second paramter of the monod function is Ks, the  "half-velocity constant", which is the number of reads needed to obtain half of the unique reads. After this point you are more likely to get reads that you already have than new unique reads. 



```{r, fitMonodCurve, echo = FALSE, fig.width = 6, fig.height = 6, fig.caption = "Plot of number of unique reads against total number of reads", message = FALSE,results='hide', include = FALSE, fig.align = "center"}

### fit monod function for each sample

library(FME)


Model <- function(p, x) return(data.frame(x = x, y = p[1]*x/(x+p[2])))
Residuals <- function(p) (Obs$y - Model(p, Obs$x)$y)


monodPars<-matrix(data = NA, ncol = 2, nrow = nrow(sampleSheet))
for(i in 1:nrow(sampleSheet)){
	Obs<-data.frame(x = hist.data$count, y= hist.data[,i+1])
	## remove NAs
	Obs<-Obs[!is.na(Obs$y),]
	P <- modFit(f = Residuals, p = c(0.1, 1))
	#sP <- summary(P)
	monodPars[i,]<-P$par
}

## plot monod function to check fit

par(mfrow = c(2, 2))
for(i in 1:nrow(sampleSheet)){
	Obs<-data.frame(x = hist.data$count, y= hist.data[,i+1])
	## remove NAs
	Obs<-Obs[!is.na(Obs$y),]
	plot(Obs, pch = 16, cex = 1.2, xlab = "nMillionReads", ylab = "nMillionUniqueReads", main = sampleSheet$SampleID[i], ylim = c(0,monodPars[i,1]), xlim = c(0,max(c(Obs$x,monodPars[i,2]), na.rm = TRUE)), axes = FALSE, cex.axis = 1.5, cex.lab = 1.5)
	lines(Model(monodPars[i,], seq(from = 0, to = max(Obs$x), by = 10000)))
	abline(h=monodPars[i,1], col = "red")
	abline(v = monodPars[i,2], col = "red")
	axis(1, at = seq(0,max(c(Obs$x,monodPars[i,2]), na.rm = TRUE), 2e7), seq(0,max(c(Obs$x,monodPars[i,2]), na.rm = TRUE), 2e7)/1000000, cex.axis = 1.5, cex.lab = 1.5)
	axis(2, at = seq(0,monodPars[i,1], 1e6), seq(0,monodPars[i,1], 1e6)/1000000, cex.axis = 1.5, cex.lab = 1.5)
	box()

}

```

Figure 2 below compares the estimated parameters for the fitted Monod functions across the samples. Library complexity is indicated by the x axis, where higher values reflect increased complexity, and sequencing sufficiency is indicated by the y axis where values > 1 mean that we are past the fast growing point and < 1 means the amount of unique reads is still growing. 

```{r, summarizeMonodFit, echo = FALSE, fig.width = 6, fig.height = 6, fig.caption = "Figure 2: Scatterplot of library complexity against sequencing efficiency",results='hide', fig.align = "center"}

#plot(monodPars/100000)

## to gauge how close to saturation we are calc ratio between nReads and Ks.
KsRatio<-r1_dat[,1]/monodPars[,2]

## plot assymptote against ratio above.
plot(monodPars[,1]/1000000, KsRatio, xlab = "Total nUnique Reads", ylab = "Ratio nReads:Ks", pch = 16, col = lineCol, cex.axis = 1.5, cex.lab = 1.5)
legend("topright", sampleSheet$SampleID, pch = 16, col = lineCol, lwd = 2, cex = 0.8)
abline(h = 1, lty = 2)
	
```


## 2. Alignment


### 2.1 Summary of alignment (pre-filtering)

The table below summarises the alignment rate and number of mapped reads per sample, before and after filtering. While general convention is a unique mapping rate over 80% is typical for a successful ATAC-seq experiment, ENCODE would require that the percentage of mapped reads, should be greater than 95%, though values >80% may be acceptable. In the table below the rows in red highlight samples with mapping rate < 80%.  

```{r, echo = FALSE, fig.width = 6, fig.height = 6}

bowtieStats<-bowtieStats[,c("Sample", "paired_total", "paired_aligned_one", "paired_aligned_multi","overall_alignment_rate")]
## calculate alignment rate based on both members of the pair aligning concordantly

bowtieStats<-cbind(bowtieStats, rowSums(bowtieStats[,3:4])/bowtieStats[,2]*100)
colnames(bowtieStats)<-c("Sample", "nPairs", "nPairsAlignOnce", "nPairsAlignMulti", "overallAlignmentRate", "pairAlignmentRate") 
bowtieStats$Sample<-gsub("_trimmed.bowtie", "", as.character(bowtieStats$Sample))


kable(bowtieStats, digits = 2) %>%
  kable_styling(full_width = F, position = "center") %>%
    row_spec(which(bowtieStats$overallAlignmentRate < 80), color = "white", background = "#D7261E")

```


### 2.1 Duplicate reads

Because of the PCR amplification, there might be read duplicates (different reads mapping to exactly the same genomic region) from overamplification of some regions. As the Tn5 insertion is random within an accessible region, we do not expect to see fragments with the same coordinates. We consider such fragments to be PCR duplicates.The histogram below shows the distribution of duplication rates based on aligned reads. 


```{r, echo = FALSE, fig.width = 5, fig.height = 5, fig.align = "center"}
hist((1-endcodeQC[,5])*100, xlab = "% duplicate", ylab = "nSamples", main = "")

```

### 2.2 Mitochondrial genome

It can be a useful QC to assess the number of mitochondrial reads; some ATAC-Seq samples have been reported to be 80% mitochondrial reads. 

```{r, echo = FALSE, fig.width = 5, fig.height = 5, fig.align = "center"}
MTpercent<-function(data){
	return(data[which(data$V1 == "chrM"),3] /sum(data[,3])*100)
}

hist(unlist(lapply(bamStats, MTpercent)), xlab = "% reads mapped to MT", ylab = "nSamples", main = "")

```


## 2.3 Library Complexity

In this section we will assess the number of unique reads as the number of total reads increases. This is equivalent to subsampling the data. The dynamics of the resulting curve indicate the complexity of the library prep and are best interpreted across a number of samples. NB this will differ slightly to the compleity curves produced earlier as this is based on just the aligned reads.


```{r, echo = FALSE, fig.width = 8, fig.height = 8, fig.align = "center"}
xlim<-range(unlist(lapply(libComplexValues, "[", "reads")))/10^6
ylim<-range(unlist(lapply(libComplexValues, "[", "values")))/10^6
plot(x=libComplexValues[[1]]$reads/10^6, y=libComplexValues[[1]]$values/10^6,  
         type="o", xlab ="Putative sequenced fragments (Million)", 
         ylab="Distinct fragments (Million)",
         main="Estimation of ATAC-seq\nlibrary complexity", col = lineCol[1], xlim = xlim, ylim = ylim)
for(i in 2:nrow(sampleSheet)){
	lines(x=libComplexValues[[i]]$reads/10^6, y=libComplexValues[[i]]$values/10^6,col = lineCol[i],type="o")
}
legend("topleft", sampleSheet$SampleID, lwd = 1.8, col = lineCol) 

```

We will also calculate additional statistics recommended by the ENCODE ATAC-seq pipeline: PCR Bottlenecking Coefficients 1 & 2 (PBC1 & PBC2) and Non-Redundant Fraction (NRF) and compare these to their reference values.

```{r,echo = FALSE, fig.width = 6, fig.height = 6, fig.align = "center"}

colnames(endcodeQC)<-c("TotalReadPairs","DistinctReadPairs","OneReadPair","TwoReadPairs","NRF","PBC1","PBC2")

hist(endcodeQC[,"NRF"], breaks = seq(min(c(endcodeQC[,"NRF"], 0.65)), max(c(endcodeQC[,"NRF"], 0.95)), length.out = 10), main = "", xlab = "Non-Redundant Fraction (NRF)", ylab = "nSamples")
abline(v = 0.7, lty = 2)
abline(v = 0.9, lty = 2)
mtext("Concerning", at = 0.65, side = 3)
mtext("Acceptable", at = 0.8, side = 3)
mtext("Ideal", at = 0.95, side = 3)

hist(endcodeQC[,"PBC1"], breaks = seq(min(c(endcodeQC[,"PBC1"], 0.65)), max(c(endcodeQC[,"PBC1"], 0.95)), length.out = 10), main = "", xlab = "PCR Bottlenecking Coefficient 1 (PBC1)", ylab = "nSamples")
abline(v = 0.7, lty = 2)
abline(v = 0.9, lty = 2)
mtext("Severe", at = 0.65, side = 3)
mtext("Moderate", at = 0.8, side = 3)
mtext("None", at = 0.95, side = 3)

hist(endcodeQC[,"PBC2"], breaks = seq(min(c(endcodeQC[,"PBC2"], 0.95)), max(c(endcodeQC[,"PBC2"], 3.05)), length.out = 10), main = "", xlab = "PCR Bottlenecking Coefficient 2 (PBC2)", ylab = "nSamples")
abline(v = 1, lty = 2)
abline(v = 3, lty = 2)
mtext("Severe", at = 0.95, side = 3)
mtext("Moderate", at = 2, side = 3)
mtext("None", at = 3.05, side = 3)

```


## 2.4 Fragment Size Distribution

ATAC-seq data should be enriched for specific sized fragments, so the distribution of fragement size should show periodicity. Typically there should be a large proportion of reads with fragment length < 100bp indicating nucleosome-free fragments, the plot show then show periodicity of ~ 200bp representing  enrichment of mono-, di-, and tri-nucleosome bound fragments (~200, 400, 600?p, respectively).  In the plot below, a sucessful ATAC-seq experiment should have a peak at ~50-100bp with a large proportion of reads less than 100 bp, and peaks at 200bp, 400bp,600bp. If there was no chromatin in the sample the graph would have no periodicity and resemble a poisson distribution with lambda = 1).

```{r,echo = FALSE, fig.width = 10, fig.height = 6, fig.align = "center"}

standardizeValues<-function(frag.len){
    x <- 1:1010
    frag.len <- frag.len[match(x, names(frag.len))]
    frag.len[is.na(frag.len)] <- 0
    y <- frag.len / sum(frag.len)
    y <- as.numeric(y)
    names(y) <- x
	return(y)
	}

tmp<-lapply(fragSizeHist,standardizeValues)
x <- 1:1010
par(mar=c(4.5, 4.5, 1, 1))
ylim<-range(tmp)*10^3
plot(x, tmp[[1]]*10^3, 
         xlim=c(0, 1010), ylim = ylim,
         xlab="Fragment length (bp)",
         ylab="Normalized  read  density (x1000)",
         type="l", col = lineCol[1])
for(i in 2:nrow(sampleSheet)){
	lines(x,tmp[[i]]*10^3, col = lineCol[i])
  }
 legend("topright", colnames(hist.data)[-1], lty = 1, col = lineCol, lwd = 2)


```

### 2.5 Cross-correlation

Normalized Strand Cross-correlation coefficient (NSC):
A measure of enrichment derived without dependence on prior determination of enriched regions. Forward and reverse strand read coverage signal tracks are computed (number of unique mapping read starts at each base in the genome on the + and - strand counted separately). The forward and reverse tracks are shifted towards and away from each other by incremental distances and for each shift, the Pearson correlation coefficient is computed. In this way, a cross-correlation profile is computed, representing the correlation between forward and reverse strand coverage at different shifts. The highest cross-correlation value is obtained at a strand shift equal to the predominant fragment length in the dataset as a result of clustering/enrichment of relative fixed-size fragments around the binding sites of the target factor or feature.

The NSC is the ratio of the maximal cross-correlation value (which occurs at strand shift equal to fragment length) divided by the background cross-correlation (minimum cross-correlation value over all possible strand shifts). Higher values indicate more enrichment, values less than 1.1 are relatively low NSC scores, and the minimum possible value is 1 (no enrichment). This score is sensitive to technical effects; for example, high-quality antibodies such as H3K4me3 and CTCF score well for all cell types and ENCODE production groups, and variation in enrichment in particular IPs is detected as stochastic variation. This score is also sensitive to biological effects; narrow marks score higher than broad marks (H3K4me3 vs H3K36me3, H3K27me3) for all cell types and ENCODE production groups, and features present in some individual cells, but not others, in a population are expected to have lower scores.

Relative Strand Cross-correlation coefficient (RSC):
A measure of enrichment derived without dependence on prior determination of enriched regions. Forward and reverse strand read coverage signal tracks are computed (number of unique mapping read starts at each base in the genome on the + and - strand counted separately). The forward and reverse tracks are shifted towards and away from each other by incremental distances and for each shift, the Pearson correlation coefficient is computed. In this way, a cross-correlation profile is computed representing the correlation values between forward and reverse strand coverage at different shifts. The highest cross-correlation value is obtained at a strand shift equal to the predominant fragment length in the dataset as a result of clustering/enrichment of relative fixed-size fragments around the binding sites of the target factor. For short-read datasets (< 100 bp reads) and large genomes with a significant number of non-uniquely mappable positions (e.g., human and mouse), a cross-correlation phantom-peak is also observed at a strand-shift equal to the read length. This read-length peak is an effect of the variable and dispersed mappability of positions across the genome. For a significantly enriched dataset, the fragment length cross-correlation peak (representing clustering of fragments around target sites) should be larger than the mappability-based read-length peak.

The RSC is the ratio of the fragment-length cross-correlation value minus the background cross-correlation value, divided by the phantom-peak cross-correlation value minus the background cross-correlation value. The minimum possible value is 0 (no signal), highly enriched experiments have values greater than 1, and values much less than 1 may indicate low quality.

## 3. Nucleosome positioning

Tn5 transposase has been shown to bind as a dimer and inserts two adaptors into accessible DNA locations separated by 9 bp. Therefore, for downstream analysis, such as peak-calling and footprinting, all reads in input bamfile need to be shifted. All reads aligning to the positive strand are offset by +4bp, and all reads aligning to the negative strand are offset by -5bp.


### 3.1 Promoter/Transcript body (PT) score

PT score is calculated as the coverage of promoter divided by the coverage of its transcript body. PT score will show if the signal is enriched in promoters. Figure X below presents a violin plot per sample, where values greater than 0 indicate that transcripts have higher signal in their promotor than the transcript. 

```{r,echo = FALSE, fig.width = 6, fig.height = 6, fig.align = "center"}
vioplot(lapply(lapply(pt.collate, mcols), "[", "PT_score"), ylab = "Promotor:Transcript Score", col = lineCol)
abline(h = 0)
```


### 3.2 Nucleosome Free Regions (NFR) score

NFR score is a ratio between cut signal adjacent to TSS and that flanking the corresponding TSS. Each TSS window of 400 bp is first divided into 3 sub-regions: the most upstream 150 bp (n1), the most downstream of 150 bp (n2), and the middle 100 bp (nf). Then the number of fragments with 5 ends overlapping each region are calculated for each TSS. The NFR score for each TSS is calculated as NFR-score = log2(nf) - log2((n1+n2)/2). Figure X below presents a violin plot per sample, where values greater than 0 indicate that signals are higher at the TSS compared to the flanking regions. 

```{r,echo = FALSE, fig.width = 6, fig.height = 6, fig.align = "center"}
vioplot(lapply(lapply(nfr.collate, mcols), "[", "NFR_score"), ylab = "Nucleosome Free Regions Score", col = lineCol)
abline(h = 0)
```


### 3.3 Transcription Start Site (TSS) Enrichment Score

The TSS enrichment calculation is a signal to noise calculation. The reads around a reference set of TSSs are collected to form an aggregate distribution of reads centered on the TSSs and extending to 1000 bp in either direction (for a total of 2000bp). This distribution is then normalized by taking the average read depth in the 100 bps at each of the end flanks of the distribution (for a total of 200bp of averaged data) and calculating a fold change at each position over that average read depth. This means that the flanks should start at 1, and if there is high read signal at transcription start sites (highly open regions of the genome) there should be an increase in signal up to a peak in the middle. We take the signal value at the center of the distribution after this normalization as our TSS enrichment metric. Based on alignment to hg38, ENCODE provide the following reference values, < 5 concerning; 5-7 acceptable; > 7 ideal.  

```{r,echo = FALSE, fig.width = 6, fig.height = 6, fig.align = "center"}



```




### 3.4 Heatmap and coverage curve for nucleosome positions

By averaging the signal across all active TSSs, we should observe that nucleosome-free fragments are enriched at the TSSs, whereas the nucleosome-bound fragments should be enriched both upstream and downstream of the active TSSs and display characteristic phasing of upstream and downstream nucleosomes. Because ATAC-seq reads are concentrated at regions of open chromatin, users should see a strong nucleosome signal at the +1 nucleosome, but the signal decreases at the +2, +3 and +4 nucleosomes. In order to visualise this reads need to be split into different bins, namely nucleosome free, mononucleosome, dinucleosome, and trinucleosome.  This is performed using a random forest to classify the fragments based on fragment length, GC content and conservation scores. The random forest is trained using the top 10% of short reads (reads below 100_bp) as nucleosome-free regions and the top 10% of intermediate length reads as (reads between 180 and 247 bp) mononucleosome. Reads that do not fit into any of the bins will be discarded.



```{r,echo = FALSE, fig.width = 6, fig.height = 6, fig.align = "center", include = FALSE}
## log2 transformed signals
library(TxDb.Hsapiens.UCSC.hg38.knownGene) ## need to loaded here rather than beginning as conflicts with dplyr
txs <- transcripts(TxDb.Hsapiens.UCSC.hg38.knownGene)
TSS <- promoters(txs, upstream=0, downstream=1)
TSS <- unique(TSS)
NTILE <- 51
dws <- ups <- 1010

logTrans<-function(list){
  return(lapply(lapply(list, "+", 1), log2))
}
sigs.log2 <- lapply(sigs.collate, logTrans)
#plot heatmap
#First, in the heatmap below, each row reprsents a TSS and the columns positions around the TSS. In the panel on the left representing the nucleosome-free fragments, there should be a hotspot (coloured red) in the centre, and in the panel on the right representing the mononucleosome fragments there should be a cold spot (coloured  yellow) in the centre and the edges of the plot hot (i.e. red).
#featureAlignedHeatmap(sigs.log2[[1]], reCenterPeaks(TSS, width=ups+dws), zeroAt=.5, n.tile=NTILE)
```

Typical TSS enrichment plot shows that nucleosome-free fragments are enriched at TSS, while mono-nucleosome fragments are depleted at TSS but enriched at flanking regions. The line graph below, plots the standardardized coverage of reads within 1kp of a TSS. The first plot includes the nucleosome free fragments, which should peak at 0bp indictaing an enrichment of fragments at the TSS. The second plot includes the mononucleosome fragments, which should have a depression at 0, and peaks of signal either side. 

```{r, include = FALSE}

## rescale the nucleosome-free and nucleosome signals to 0~1
out.nucleosome<-matrix(data = NA, ncol = nrow(sampleSheet), nrow = NTILE)
out.mono<-matrix(data = NA, ncol = nrow(sampleSheet), nrow = NTILE)
for(i in 1:nrow(sampleSheet)){
  out <- featureAlignedDistribution(sigs.collate[[i]], 
                                  reCenterPeaks(TSS, width=ups+dws),
                                  zeroAt=.5, n.tile=NTILE, type="l", 
                                ylab="Averaged coverage")
   out.nucleosome[,i]<-out[,"NucleosomeFree"]
    out.mono[,i]<-out[,"mononucleosome"] 
}

```

```{r,echo = FALSE, fig.width = 6, fig.height = 6, fig.align = "center"}
range01 <- function(x){(x-min(x))/(max(x)-min(x))}
## standardize
out.mono <- apply(out.mono, 2, range01)
out.nucleosome <- apply(out.nucleosome, 2, range01)
plot(out.nucleosome[,1], type="l", xaxt="n", 
        xlab="Position (bp)", 
        ylab="Fraction of signal", col = lineCol[1], main = "Nucleosome Free")
for(i in 2:nrow(sampleSheet)){
  lines(out.nucleosome[,i], col = lineCol[i])
}
axis(1, at=seq(0, 50, by=5)+1, 
     labels=c("-1K", seq(-800, 800, by=200), "1K"), las=2)
abline(v=seq(0, 100, by=10)+1, lty=2, col="gray")

plot(out.mono[,1], type="l", xaxt="n", 
        xlab="Position (bp)", 
        ylab="Fraction of signal", col = lineCol[1], main = "Mononucleosome")
for(i in 2:nrow(sampleSheet)){
  lines(out.mono[,i], col = lineCol[i])
}
axis(1, at=seq(0, 50, by=5)+1, 
     labels=c("-1K", seq(-800, 800, by=200), "1K"), las=2)
abline(v=seq(0, 100, by=10)+1, lty=2, col="gray")


```


### 3.5 Relative enrichment in genomic fetaures

This section uses genomic annotation to show where reads map in terms of genomic features. Below this is represented as a heatmap showing the enrichment of reads compared to the background levels of the feature. Typically we would expect more than half of the peaks fall into enhancer regions (distal intergenic and intronic regions), and only around 25% of the peaks are in promoter regions.


```{r, plotGenomicHeatmap, echo = FALSE, fig.width = 8, fig.height = 6, fig.cap="Figure X: Heatmap genomic feature enrichment",results='hide'}
plotRegi(dat, facet=FALSE)
```


## 4. Peak calling

## 4.1 Summary of peaks

Below is a histogram showing the distribution of the number of peaks per sample. 

```{r,echo = FALSE, fig.width = 5, fig.height = 5, fig.align = "center"}
hist(unlist(lapply(peaks(dat), length)))
```

Do we see a relationship between number of aligned reads and peaks. If samples have comparable library complexity (both biologically and technically) we would expect a consisent relathionship across the samples.

```{r, scatterplotReadsPeaks,echo = FALSE, fig.width = 7, fig.height = 7, fig.align = "center"}

plot(reads(dat)/1000000, unlist(lapply(peaks(dat), length)), pch = 16, xlab = "nMillionReads", ylab = "nPeaks", col = lineCol)
legend("topright", sampleSheet$SampleID, pch = 16, col = lineCol)

```




## 4.1 Plot Peak Profile

Figure X represents the mean read depth across and around peaks centered
on the summit (point of highest pileup) for each peak, with each sample presented by a different coloured line. Good ChIPs will show distinctive patterns of enrichment in these peaks, while associated controls will be relatively flat. These profiles can vary depending on what type of mark is being studied b  transcription factor, histone mark, or other DNAbinding protein such as a polymerase b  but similar marks usually have a distinctive profile in successful ChIPs. This plot can be used to identify different peak shapes as well differences in absolute heights which can be helpful for identifing outliers. 

```{r, echo = FALSE, fig.width = 8, fig.height = 6, fig.cap="Figure X: Plot of the average signal profile across peaks",results='hide'}
#plotPeakProfile(dat, facet = FALSE)
peakDat<-Normalisedaveragepeaksignal(dat)

y_lim<-range(peakDat, na.rm = TRUE)
plot(peakDat[,1], xlab = "Distance", ylab = "Normalised Signal", col = lineCol[1], type = "l", lwd = 1.5, ylim = y_lim, axes = FALSE)
for(i in 2:ncol(peakDat)){
  lines(peakDat[,i], col = lineCol[i], lwd = 1.5)
}
axis(1, at = seq(0,400, 50), seq(-200, 200, 50))
axis(2)
box()
legend("topright", sampleSheet$SampleID, pch = 16, col = lineCol, lwd = 2, cex = 0.8)

```


## 4.2 Specificity of reads in peaks

Boxplot of the distribution of reads in peaks. 

```{r, echo = FALSE, fig.width = 10, fig.height = 6, fig.cap="Figure X: Barplot of the proportion of reads in called peaks"}
plotRap(dat, facet = FALSE)

```

Barplot of the percentage of reads in peaks called for that sample. This plot gives a good indication of how "enriched" the sample is and can be considered a "signal-to-noise" measure of what proportion of the library consists of fragments from binding sites vs. background reads. According to the ENCODE standard processing pipeline for ATAC-seq the FRiP score should be > 0.3, though values greater than 0.2 are acceptable. 

```{r, plotFrip, echo = FALSE, fig.width = 5, fig.height = 5, fig.cap="Figure X: Barplot of the % fragments in peaks per sample"}
y_lim<-c(0,max(frip(dat)*100, na.rm = TRUE)*1.1)
barplot(frip(dat)*100, col = lineCol, ylim = y_lim, ylab = "% Reads in Peaks", names.arg = sampleSheet$IID, cex.axis = 1.5, cex.lab = 1.5)
legend("topright", sampleSheet$SampleID, pch = 16, col = lineCol, lwd = 2, cex = 0.8)
abline(h = 0.2, lty = 2)
abline(h = 0.3, lty = 2)
```

## 4.3 Enrichment in genomic features

Here we will consider a more extended set of genomic annotations and look at the distribution of genomic locations of peaks. 

```{r,plotGenomicAnnotation, echo = FALSE, fig.width = 8, fig.height = 6, fig.cap="Barplot of genomic feature annotation", message = FALSE, warning = FALSE,results='hide'}
library(clusterProfiler)
## only works for samples with > 0 peaks
nPeaks<-unlist(lapply(peaks(dat), length))
peakAnnoList <- lapply(peaks(dat)[which(nPeaks > 0)], annotatePeak,tssRegion=c(-500, 500), 
                         TxDb=TxDb.Hsapiens.UCSC.hg38.knownGene, 
                         annoDb="org.Hs.eg.db")
plotAnnoBar(peakAnnoList)

```

Third, we will look at the proximity of peaks to transcription start sites.


```{r,plotTSSproximity, echo = FALSE, fig.width = 8, fig.height = 6, fig.cap="Barplot of proximity to TSS", message = FALSE, warning = FALSE,results='hide'}
plotDistToTSS(peakAnnoList)

```


### Which genes are these peaks annotated to?

```{r, pathwayAnalysis,echo = FALSE, fig.width = 8, fig.height = 6, fig.cap="",results="hide"}

genes = lapply(peakAnnoList, function(i) as.data.frame(i)$geneId)
names(genes) = sub("_", "\n", names(genes))

geneCount<-lapply(lapply(genes, table), sort, decreasing = TRUE)

allGenes<-unique(unlist(lapply(geneCount,names)))
histGeneCount<-matrix(data = 0, ncol = length(geneCount), nrow = length(allGenes))
rownames(histGeneCount)<-allGenes
for(i in 1:length(geneCount)){
	histGeneCount[names(geneCount[[i]]),i]<-geneCount[[i]]
}
rownames(histGeneCount)<-mapIds(org.Hs.eg.db, allGenes,"SYMBOL", "ENTREZID")
## find genes most frequent across the sample


totGene<-rowSums(histGeneCount)
barplot(t(histGeneCount[order(totGene, decreasing = TRUE)[1:20],]), col = lineCol, names.arg = rep("", 20), ylab = "nPeaks")

text(x = seq(from = 1, by = 1.2, length.out = 20), y = -0.5,  names(totGene)[order(totGene, decreasing = TRUE)[1:20]],srt = 45, xpd = TRUE, adj = 1)
legend("topright", sampleSheet$SampleID, pch = 15, col = lineCol)

```


## Pathway analysis of annotated genes

```{r, pathwayAnalysis2,echo = FALSE, fig.width = 6, fig.height = 6, fig.cap="",results='hide'}


compKEGG <- compareCluster(geneCluster   = genes,
                         fun           = "enrichKEGG",
                         pvalueCutoff  = 0.05,
                         pAdjustMethod = "BH")
dotplot(compKEGG, showCategory = 15, title = "KEGG Pathway Enrichment Analysis")
```


```{r, summaryDataTable, results = "asis", echo = FALSE, message = FALSE, include = FALSE}
sumTab<-cbind(sampleSheet[,c("SampleID", techVars)], reads(dat), unlist(lapply(peaks(dat), length)))
colnames(sumTab)<-c("SampleID", techVars, "AlignedReads", "Peaks")
kable(sumTab, caption = "Table 1: Summary of aligned reads and number of peaks")
```




## 5. Sample clustering

Below is a correlation heatmap is based on correlation values for all the peak scores for each sample to demonstrate how samples cluster together.

```{r,plotCorHeatmap, echo = FALSE, fig.width = 6, fig.height = 6, fig.cap="Heatmap of sample correlations",results='hide'}

plotCorHeatmap(dat)

```

Below is a plot of the first two principal component which demonstrates how samples cluster together.

```{r,plotPrincomp, echo = FALSE, fig.width = 6, fig.height = 6, fig.cap="",results='hide'}

plotPrincomp(dat, dotSize=2)

```
